silence/silence.go

```diff
@@ -662,7 +701,28 @@ func (s *Silences) expire(id string) error {
 		sil.EndsAt = now
 	}
 
-	return s.setSilence(sil, now)
+	// This code has been copied from setSilence to avoid an issue where
```

Comment:
There's a lot of risk with drift with copying this code, and I would rather avoid it if possible. Can you elaborate on your decision to outright remove the validation? What kind of risks does removing the validation pose?

I suggest you add a boolean to the arguments of `validateSilence` to skip the validation should you wish to continue with this approach.

Reply:
> The validation was removed to fix this issue https://github.com/prometheus/alertmanager/pull/3567#discussion_r1392961898

Sorry, I don't think that was my question, I read your comment and your decision-making is still not clear - It was more around the decision to remove the validation as part of solving the issue, which I guess your second comment touches a bit on.

I think I follow your line of thought, it seems strange that expiring a silence validates the silence itself. I guess it's a byproduct (assuming original author's intention) of re-using the `setSilence` function and not because the actual validation is needed. 

If we go with that assumption, please make sure is documented at the call site when you add the boolean.

Reply:
> I think I follow your line of thought, it seems strange that expiring a silence validates the silence itself. I guess it's a byproduct (assuming original author's intention) of re-using the setSilence function and not because the actual validation is needed.

Yes! I think the reason `expire` uses `setSilence` is a.) it was convenient to re-use `setSilence` for expiration and b.) once a silence has been validated in the past then, unless the silence is updated, it should always be valid in the future. Not an unreasonable assumption if validation rules never change.

However, in my opinion deletion should not do validation as otherwise we risk locking users out of specific operations (i.e. what happens if a user somehow has an invalid silence that needs to be deleted? How can the user recover from an inconsistent state to a consistent state?).